src/openvic-simulation/misc/Define.cpp

```diff
@@ -1,109 +1,249 @@
+}
+
+bool DefineManager::load_define_years(Timespan& value, Define::Type type, std::string_view name) const {
+	return _load_define_timespan<Timespan::from_years>(value, type, name);
+}
+
+DefineManager::DefineManager()
+  : // Date
+	start_date { 1836, 1, 1 },
+	end_date { 1936, 1, 1 }
+
+	// Country
+
+	// Economy
+
+	// Military
+
+	// Diplomacy
+
+	// Pops
+
+	// Ai
+
+	// Graphics
+
+	{}
+
+bool DefineManager::add_define(std::string_view name, std::string_view value, Define::Type type) {
+	if (name.empty()) {
+		Logger::error("Invalid define identifier - empty!");
+		return false;
+	}
+
+	if (value.empty()) {
+		Logger::error("Invalid define value for \"", name, "\" - empty!");
 		return false;
 	}
-	return defines.add_item({ name, date.to_string(), Define::Type::Date });
+
+	return defines.add_item({ name, value, type }, duplicate_warning_callback);
+}
+
+bool DefineManager::in_game_period(Date date) const {
+	return date.in_range(start_date, end_date);
 }
 
 bool DefineManager::load_defines_file(ast::NodeCPtr root) {
+	using enum Define::Type;
+
 	bool ret = expect_dictionary_keys(
 		"defines", ONE_EXACTLY, expect_dictionary([this](std::string_view key, ast::NodeCPtr value) -> bool {
-			using enum Define::Type;
-			static const string_map_t<Define::Type> type_map {
-				{ "country",   Country },
-				{ "economy",   Economy },
-				{ "military",  Military },
-				{ "diplomacy", Diplomacy },
-				{ "pops",      Pops },
-				{ "ai",        Ai },
-				{ "graphics",  Graphics },
-			};
 
-			const string_map_t<Define::Type>::const_iterator type_it = type_map.find(key);
+			const Define::Type type = Define::string_to_type(key);
 
-			if (type_it != type_map.end()) {
+			if (type != Unknown) {
 
 				return expect_dictionary_reserve_length(
 					defines,
-					[this, &key, type = type_it->second](std::string_view inner_key, ast::NodeCPtr value) -> bool {
-						std::string str_val;
-						bool ret = expect_identifier_or_string(assign_variable_callback_string(str_val))(value);
-						ret &= add_define(inner_key, std::move(str_val), type);
-						return ret;
+					[this, type](std::string_view inner_key, ast::NodeCPtr value) -> bool {
+						return expect_identifier_or_string(
+							[this, &inner_key, type](std::string_view value) -> bool {
+								return add_define(inner_key, value, type);
+							}
+						)(value);
 					}
 				)(value);
 
 			} else if (key == "start_date" || key == "end_date") {
 
-				return expect_date_identifier_or_string(std::bind_front(&DefineManager::add_date_define, this, key))(value);
+				return expect_identifier_or_string(
+					[this, &key](std::string_view value) -> bool {
+						return add_define(key, value, Date);
+					}
+				)(value);
 
 			} else {
+
+				Logger::error("Invalid define type - \"", key, "\"");
 				return false;
+
 			}
 		})
 	)(root);
 
 	lock_defines();
 
+	// Date
+	ret &= load_define(start_date, Date, "start_date");
+	ret &= load_define(end_date, Date, "end_date");
```

Comment:
Obviously this is both sensible and necessary: if you are going to be using the define during game play, you don't want to be doing a string lookup. But, doesn't this apply to *all* of the defines? And if it applies to all of them, such that you will eventually have one variable per define value, then you don't need the string map -> defines at all. You can just load them into their variables and omit the complexity of the map completely.

Reply:
Yeah I'm considering doing away with `DefineManager` altogether and directly loading the values into member variables of other appropriate managers, both to remove the unnecessary intermediate storage and to avoid needing to ensure simulation functions have access to yet another manager. I wasn't the original author or reviewer of this approach, and although I'm gonna stick with adaptating it and getting the work I've built on top of it merged for now, after that I'll look into a better repo-wide approach to constants and include the `defines.lua` ones in that.